sql/core/src/test/scala/org/apache/spark/sql/execution/BaseScriptTransformationSuite.scala

```diff
@@ -0,0 +1,296 @@
+    case class Struct(d: Int, str: String)
+    withTempView("v") {
+      val df = Seq(
+        (1, "1", 1.0, 11.toByte, BigDecimal(1.0), new Timestamp(1),
+          new Date(2020, 7, 1), new CalendarInterval(7, 1, 1000), Array(0, 1, 2),
+          Map("a" -> 1)),
+        (2, "2", 2.0, 22.toByte, BigDecimal(2.0), new Timestamp(2),
+          new Date(2020, 7, 2), new CalendarInterval(7, 2, 2000), Array(3, 4, 5),
+          Map("b" -> 2)),
+        (3, "3", 3.0, 33.toByte, BigDecimal(3.0), new Timestamp(3),
+          new Date(2020, 7, 3), new CalendarInterval(7, 3, 3000), Array(6, 7, 8),
+          Map("c" -> 3))
+      ).toDF("a", "b", "c", "d", "e", "f", "g", "h", "i", "j")
+        .select('a, 'b, 'c, 'd, 'e, 'f, 'g, 'h, 'i, 'j, struct('a, 'b).as("k"))
+      // Note column d's data type is Decimal(38, 18)
+      df.createTempView("v")
+
+      assert(spark.table("v").schema ==
+        StructType(Seq(StructField("a", IntegerType, false),
+          StructField("b", StringType, true),
+          StructField("c", DoubleType, false),
+          StructField("d", ByteType, false),
+          StructField("e", DecimalType(38, 18), true),
+          StructField("f", TimestampType, true),
+          StructField("g", DateType, true),
+          StructField("h", CalendarIntervalType, true),
+          StructField("i", ArrayType(IntegerType, false), true),
+          StructField("j", MapType(StringType, IntegerType, false), true),
+          StructField("k", StructType(
+            Seq(StructField("a", IntegerType, false),
+              StructField("b", StringType, true))), false))))
+
+      // Can't support convert script output data to ArrayType/MapType/StructType now,
+      // return these column still as string
+      checkAnswer(
+        df,
+        (child: SparkPlan) => createScriptTransformationExec(
+          input = Seq(
+            df.col("a").expr,
+            df.col("b").expr,
+            df.col("c").expr,
+            df.col("d").expr,
+            df.col("e").expr,
+            df.col("f").expr,
+            df.col("g").expr,
+            df.col("h").expr,
+            df.col("i").expr,
+            df.col("j").expr,
+            df.col("k").expr),
+          script = "cat",
+          output = Seq(
+            AttributeReference("a", IntegerType)(),
+            AttributeReference("b", StringType)(),
+            AttributeReference("c", DoubleType)(),
+            AttributeReference("d", ByteType)(),
+            AttributeReference("e", DecimalType(38, 18))(),
+            AttributeReference("f", TimestampType)(),
+            AttributeReference("g", DateType)(),
+            AttributeReference("h", CalendarIntervalType)(),
+            AttributeReference("i", StringType)(),
+            AttributeReference("j", StringType)(),
+            AttributeReference("k", StringType)()),
+          child = child,
+          ioschema = defaultIOSchema
+        ),
+        df.select(
+          'a, 'b, 'c, 'd, 'e, 'f, 'g, 'h,
```

Comment:
What happens if `cat` output an incompatible string against an output data type? What's a current behaivour of hive TRANSFORM and this PR?

Reply:
> What happens if `cat` output an incompatible string against an output data type? What's a current behaivour of hive TRANSFORM and this PR?

In current spark 3.x, the default serde is `LazySimpleSerde`. If return string is incompatible, it will return `null`
```
  test("test string incompatible") {
    sql(
      """
        |select transform(a, b, c)
        |using 'cat' as (a string, b int , c int)
        |from (select 1 as a, "a" as b, cast(2000 as timestamp) as c) tmp
      """.stripMargin).show()
  }
```

Hive default serde
```
[info] - script transformation should not swallow errors from upstream operators (no serde) (1 second, 248 milliseconds)
+---+----+----+
|  a|   b|   c|
+---+----+----+
|  1|null|null|
+---+----+----+
```


Spark default format
```
[info] - test string incompatible *** FAILED *** (553 milliseconds)
[info]   org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 4.0 failed 1 times, most recent failure: Lost task 0.0 in stage 4.0 (TID 8, 192.168.31.129, executor driver): java.lang.NumberFormatException: For input string: "a"
```
For  me, to show the error is better since we can know the data error and know the error quickly.
Do you think we need to catch the error and return null?

Reply:
> What happens if `cat` output an incompatible string against an output data type? What's a current behaivour of hive TRANSFORM and this PR?

For CalenderIntervalType
```
 test("test string incompatible") {
    sql(
      """
        |select transform(a, b, c, d)
        |using 'cat' as (a string, b int , c int, d interval)
        |from
        |(
        |select
        | 1 as a,
        | "a" as b,
        | cast(2000 as timestamp) as c,
        | CAST('interval 3 month 1 hour' AS interval) as d
        |) tmp
      """.stripMargin).show()
  }
```


Hive throw error 
```
[info] - test string incompatible *** FAILED *** (460 milliseconds)
[info]   org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 4.0 failed 1 times, most recent failure: Lost task 0.0 in stage 4.0 (TID 4, 192.168.31.129, executor driver): scala.MatchError: CalendarIntervalType (of class org.apache.spark.sql.types.CalendarIntervalType$)
[info] 	at org.apache.spark.sql.hive.HiveInspectors$typeInfoConversions.toTypeInfo(HiveInspectors.scala:1044)
[info] 	at org.apache.spark.sql.hive.execution.HiveScriptTransformationExec.$anonfun$initSerDe$1(HiveScriptTransformationExec.scala:95)
[info] 	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)

```

Spark now return 
```
[info] - script transformation should not swallow errors from upstream operators (no serde) (1 second, 272 milliseconds)
+---+---+-------------------+----------------+
|  a|  b|                  c|               d|
+---+---+-------------------+----------------+
|  1|  a|1969-12-31 16:33:20|3 months 1 hours|
+---+---+-------------------+----------------+
```

Reply:
Thanks for the check. hm, I personally think our Spark default serde also would better to follows the `LazySimpleSerde` behaivour so that users don't get suprised at the behaivour difference between them. I think most built-in functions also return null for irregal input.

Reply:
Either way, could you add tests for that case?

Reply:
> Either way, could you add tests for that case?

Updated.